Add #country_regions endpoint

RadekMolenda opened the pull request:
This endpoint will allow users to fetch all the administrative regions in one 
request. This sometimes might be quite useful

Commit:
Change parameter name to mirror geonames API

Commit:
Add #country_regions endpoint

ChuckJHardy commented:
:+1:

mfilej commented:
I have figured out why we can rename fcode! Give me a few minutes.

mfilej commented:
Ok so we said we want to keep `fcode` because that is what is returned by the api. But! We have this layer where we get do dictate the results as well: https://github.com/Leadformance/gemonames/blob/master/lib/gemonames/api_client.rb#L135-L143

So if in this layer we make sure that we return an attribute with a name that actually makes sense, we can use the same attribute name as input.

I think this is important in this case because `fcode` means nothing and it is not even documented in their docs. I think using `admin_code` instead would make a lot of sense.

Wdyt?

(If you don't want to deal with it I'm happy to do the necessary changes). :smile:

mfilej committed:
Extract result classes to please codeclimate limit

mfilej commented:
@RadekMolenda Updated to please CodeClimate. I thought memoizing was overkill for the null objects so I just wrapped them in lambdas. Curious to hear your thoughts.

RadekMolenda commented:
:+1:

mfilej commented on lib/gemonames/results.rb:
```diff
@@ -0,0 +1,77 @@
+    alias_method :result?, :result
+    alias_method :present?, :result
+
+    def self.with_nil
```
I wasn't sure about `with_nil` vs `with_nils`. Both sound off to be honest, but I like the symmetry with `with` which is what we use for non-null objects.

mfilej replied:
I also considered plain `null`.

mfilej committed:
Add support for feature_code attribute

This changes the search result response to include said attribute and
also changes the input parameter name from fcode to feature_code.
